component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,343 @@
+
+        return getPriority(messageContext, 24);
+    }
+
+    @Override
+    public boolean canHandle(MessageContext messageContext) {
+
+        return AuthConfigurationUtil.isAuthHeaderMatch(messageContext, DPoPConstants.OAUTH_DPOP_HEADER) ||
+                AuthConfigurationUtil.isAuthHeaderMatch(messageContext, DPoPConstants.OAUTH_HEADER);
+    }
+
+    private AuthenticationResult getAuthenticationResult(AuthenticationResult authenticationResult,
+                                                         OAuth2TokenValidationResponseDTO responseDTO,
+                                                         String authorizationHeader,
+                                                         AuthenticationRequest authenticationRequest,
+                                                         MessageContext messageContext, String accessToken,
+                                                         String consumerKey)
+            throws AuthenticationFailException {
+
+        if (!responseDTO.isValid()) {
+            log.debug(responseDTO.getErrorMsg());
+            return authenticationResult;
+        }
+
+        TokenBinding binding = responseDTO.getTokenBinding();
+        if (binding != null && DPoPConstants.OAUTH_DPOP_HEADER.equals(binding.getBindingType())) {
+            if (!authorizationHeader.startsWith(DPoPConstants.OAUTH_DPOP_HEADER)) {
+                log.debug("DPoP prefix is not defined correctly in the Authorization header.");
+                return authenticationResult;
+            }
+            String dpopHeader = authenticationRequest.getHeader(DPoPConstants.OAUTH_DPOP_HEADER);
+
+            if (StringUtils.isBlank(dpopHeader)) {
+                log.debug("DPoP header is empty.");
+                return authenticationResult;
+            }
+            try {
+                String thumbprintOfPublicKey = Utils.getThumbprintOfKeyFromDpopProof(dpopHeader);
+                if (StringUtils.isBlank(thumbprintOfPublicKey)) {
+                    log.debug("Thumbprint value of the public key is empty in the DPoP Proof.");
+                    return authenticationResult;
+                }
+                if (!thumbprintOfPublicKey.equalsIgnoreCase(responseDTO.getTokenBinding().getBindingValue())) {
+                    log.debug("Thumbprint value of the public key in the DPoP proof is not equal to binding value" +
+                            " of the responseDTO.");
+                    return authenticationResult;
+                }
+            } catch (IdentityOAuth2Exception e) {
+                String errorMessage = "Error occurred while getting the thumbprint of the public key from the DPoP " +
+                        "proof.";
+                log.error(errorMessage);
+                throw new AuthenticationFailException(errorMessage);
+            }
+        } else {
+            if (!authorizationHeader.startsWith(DPoPConstants.OAUTH_HEADER)) {
```

Comment:
We don;t need to handle oauth header in this handler

Reply:
Added to handle "Invoking protected API  using Bearer token with DPoP prefix in the Authorization Header".

Reply:
Fixed